modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/builtin/BaseBuiltInJSModule.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.js.loader.modules.extender.internal.npm.builtin;
+
+import com.liferay.frontend.js.loader.modules.extender.npm.JSModule;
+import com.liferay.frontend.js.loader.modules.extender.npm.JSPackage;
+import com.liferay.frontend.js.loader.modules.extender.npm.ModuleNameUtil;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Provides an imcomplete implementation of {@link JSModule} to hold modules
+ * to be served through {@link BuiltInJSModuleServlet} and
+ * {@link BuiltInJSResolvedModuleServlet}.
+ *
+ * @author Iván Zaera
+ */
+public abstract class BaseBuiltInJSModule implements JSModule {
+
+	/**
+	 * Constructs a <code>BaseBuiltInJSModule</code> with the module's package,
+	 * name, and dependencies.
+	 *
+	 * @param jsPackage the module's package
+	 * @param name the module's name
+	 * @param dependencies the module names this module depends on
+	 */
+	public BaseBuiltInJSModule(
+		JSPackage jsPackage, String name, Collection<String> dependencies,
+		JSONObject flagsJSONObject) {
+
+		_jsPackage = jsPackage;
+		_name = name;
+		_dependencies = dependencies;
+		_flagsJSONObject = flagsJSONObject;
+	}
+
+	@Override
+	public Collection<String> getDependencies() {
+		return _dependencies;
+	}
+
+	@Override
+	public Collection<String> getDependencyPackageNames() {
+		List<String> dependencyPackageNames = new ArrayList<>();
+
+		for (String dependency : _dependencies) {
+			String packageName = ModuleNameUtil.getPackageName(dependency);
+
+			if (packageName != null) {
+				dependencyPackageNames.add(packageName);
+			}
+		}
+
+		return dependencyPackageNames;
+	}
+
+	@Override
+	public JSONObject getFlagsJSONObject() {
+		return _flagsJSONObject;
+	}
+
+	@Override
+	public String getId() {
+		return ModuleNameUtil.getModuleId(_jsPackage, _name);
+	}
+
+	@Override
+	public JSPackage getJSPackage() {
+		return _jsPackage;
+	}
+
+	@Override
+	public String getName() {
+		return _name;
+	}
+
+	@Override
+	public String getResolvedId() {
+		StringBundler sb = new StringBundler(5);
+
+		sb.append(_jsPackage.getName());
+		sb.append(StringPool.AT);
+		sb.append(_jsPackage.getVersion());
+		sb.append(StringPool.SLASH);
+		sb.append(_name);
+
+		return sb.toString();
+	}
+
+	@Override
+	public String getResolvedURL() {
+		StringBundler sb = new StringBundler(2);
+
+		sb.append("/o/js/resolved-module/");
+		sb.append(getResolvedId());
+
+		return sb.toString();
+	}
+
+	@Override
+	public String getURL() {
+		StringBundler sb = new StringBundler(2);
+
+		sb.append("/o/js/module/");
```

Comment:
I forget what we concluded about "/o/" being potentially overridable. Can we avoid hardcoding these?

Reply:
It is hardcoded in `Portal.PATH_MODULE` so I guess this is safe for now. I remember someone said it can be configured, but if some day it changes, we will need to recheck the whole portal, because `Portal.PATH_MODULE` is a constant, not a method, and it can even potentially inlined in each compiled Java class (depending on the compiler used).

So, to avoid potential errors, I will leave it like it is for now...

Reply:
Note that there's a `PortalUtil.getPathModule()` too, which is a function and, this time, could be used for this purpose. However, it prepends the webapp context to `/o` so it would probably break this code if used here :shrug: .